packages/rule-sri/src/sri.ts

```diff
@@ -0,0 +1,305 @@
+enum algorithms {
+    sha256 = 1,
+    sha384 = 2,
+    sha512 = 3
+}
+
+export default class SRIRule implements IRule {
+
+    public static readonly meta: RuleMetadata = {
+        docs: {
+            category: Category.security,
+            description: `Require scripts and link elements to use Subresource Integrity`
+        },
+        id: 'sri',
+        schema: [{
+            additionalProperties: false,
+            properties: {
+                baseline: {
+                    oneOf: [Object.keys(algorithms)],
+                    type: 'string'
+                }
+            }
+        }],
+        scope: RuleScope.any
+    }
+
+    private resources: Map<string, string>;
+    private context: RuleContext;
+    private origin: string;
+    private baseline: string = 'sha384';
+
+    /**
+     * Returns the hash of the content for the given `sha` strengh in a format
+     * valid with SRI:
+     * * base64
+     * * `sha384-hash`
+     */
+    private calculateHash(content: string, sha): string {
+        const hash = crypto
+            .createHash(sha)
+            .update(content)
+            .digest('base64');
+
+        return hash;
+    }
+
+    /**
+     * Checks if the element that originated the request/response is a script or a stylesheet.
+     * There could be other downloads from a `link` element that are not stylesheets and should
+     * be ignored.
+     */
+    private isScriptOrLink(evt: FetchEnd): Promise<boolean> {
+        debug('Is <script> or <link>?');
+        const { element } = evt;
+
+        /*
+         * We subscribe to `fetch::end::script|css`, so element should always exist.
+         * "that should never happen" is the fastest way to make it happen so better be safe
+         */
+        /* istanbul ignore if */
+        if (!element) {
+            return Promise.resolve(false);
+        }
+
+        /*
+         * The element is not one that we care about (could be an img, video, etc.)
+         * No need to report anything, but we can stop processing things right away.
+         */
+        const isScript: boolean = element.nodeName === 'SCRIPT' && !!element.getAttribute('src');
+        const isStyle: boolean = element.nodeName === 'LINK' && element.getAttribute('rel') === 'stylesheet';
+
+        return Promise.resolve(isScript || isStyle);
+    }
+
+    /**
+     * Verifies if the response is elegible for integrity validation. I.E.:
+     *
+     * * `same-origin`
+     * * `cross-origin` on a CORS-enabled request
+     *
+     * More info in https://w3c.github.io/webappsec-subresource-integrity/#is-response-eligible
+     */
+    private async isEligibleForIntegrityValidation(evt: FetchEnd): Promise<boolean> {
+        debug('Is eligible for integrity validation?');
+
+        const { element, resource } = evt;
+        const resourceOrigin: string = new URL(resource).origin;
```

Comment:
Maybe use [same-origin](https://github.com/sonarwhal/sonarwhal/blob/master/packages/rule-disown-opener/package.json#L12) as things may be more complex then this?

Reply:
I actually think this code is better than [the code in that package](https://github.com/rexxars/same-origin/blob/master/index.js). It's using the old `url.parse` and doing some magic, and we get for free the `origin` if we use `new URL`. Maybe we could add a method in `misc` so we can remove that dependency from the project.

Reply:
> Maybe we could add a method in misc so we can remove that dependency from the project.

:+1:

Reply:
I've created https://github.com/sonarwhal/sonarwhal/issues/879 for this